h2/src/main/org/h2/util/ToDateTokenizer.java

```diff
@@ -0,0 +1,548 @@
+        , AM_PM(PARSLET_TIME, PATTERN_AM_PM) // Meridian indicator
+        , EE(PARSLET_YEAR) // NOT supported yet - Full era name (Japanese Imperial, ROC Official, and Thai Buddha calendars).
+        , E(PARSLET_YEAR) // NOT supported yet - Abbreviated era name (Japanese Imperial, ROC Official, and Thai Buddha calendars).
+        , Y(PARSLET_YEAR) //
+        , I(PARSLET_YEAR) //
+        , Q(PARSLET_MONTH) // Quarter of year (1, 2, 3, 4; JAN-MAR = 1).
+        //, W(PARSLET_Week) // Week of month (1-5)
+        , D(PARSLET_DAY) // Day of week (1-7).
+        , J(PARSLET_DAY); // NOT supported yet - Julian day; the number of days since Jan 1, 4712 BC.
+
+
+        private final static Map<Character, List<FormatTokenEnum>> cache = new HashMap<Character, List<FormatTokenEnum>>(FormatTokenEnum.values().length);
+        private final ToDateParslet toDateParslet;
+        private final Pattern patternToUse;
+
+        FormatTokenEnum(final ToDateParslet toDateParslet, final Pattern patternToUse) {
+            this.toDateParslet = toDateParslet;
+            this.patternToUse = patternToUse;
+        }
+
+        FormatTokenEnum(final ToDateParslet toDateParslet) {
+            this.toDateParslet = toDateParslet;
+            patternToUse = Pattern.compile(format("^(%s)", name()), Pattern.CASE_INSENSITIVE);
+        }
+
+        private static List<FormatTokenEnum> EMPTY_LIST = new ArrayList<FormatTokenEnum>(0);
+
+        /**
+         * OPTIMISATION: Only return a list of {@link FormatTokenEnum} that share the same 1st char
+         * using the 1st char of the 'to parse' formatStr. Or return empty list if no match.
+         */
+        static List<FormatTokenEnum> getTokensInQuestion(final ToDateParams params) {
+            List<FormatTokenEnum> result = EMPTY_LIST;
+            if (cache.size() <= 0) {
+                initCache();
+            }
+            String formatStr = params.getFormatStr();
+            if (formatStr != null && formatStr.length() > 0) {
+                Character key = Character.toUpperCase(formatStr.charAt(0));
+                result = cache.get(key);
+            }
+            if (result == null) {
+                result = EMPTY_LIST;
+            }
+            return result;
+        }
+
+        private static synchronized void initCache() {
+            if (cache.size() <= 0) {
+                for (FormatTokenEnum token : FormatTokenEnum.values()) {
+
+                    List<Character> tokenKeys = new ArrayList<Character>();
+
+                    if(token.name().contains("_")) {
+                       String[] tokens = token.name().split("_");
+                       for(String tokenLets : tokens) {
+                           tokenKeys.add(tokenLets.toUpperCase().charAt(0));
+                       }
+                    } else {
+                       tokenKeys.add(token.name().toUpperCase().charAt(0));
+                    }
```

Comment:
maybe just split the BC_AD and AM_PM parsing into separate enums, will simplify the patterns and this code